django/db/migrations/graph.py

```diff
@@ -207,8 +221,18 @@ def remove_replacement_node(self, replacement, replaced):
         nodes are not remapped as they are expected to be correct already.
         """
         self.nodes.pop(replacement, None)
-        # Pop without `None` here as the replacement node must exist.
-        replacement_node = self.node_map.pop(replacement)
+        try:
+            replacement_node = self.node_map.pop(replacement)
+        except KeyError as exc:
+            exc_value = NodeNotFoundError(
+                "Unable to remove replacement node %r. It was either never added"
+                " to the migration graph, or has been removed already." % (replacement, ),
+                replacement
+            )
+            exc_value.__cause__ = exc
+            if not hasattr(exc, '__traceback__'):
+                exc.__traceback__ = sys.exc_info()[2]
+            six.reraise(NodeNotFoundError, exc_value, sys.exc_info()[2])
```

Comment:
Looks like you picket that from the `_reraise_missing_dependency()` function in the migration loader. I'm not too firm with Python 2/3 traceback handling. I introduced that pattern in https://github.com/django/django/pull/3280/files#diff-a0eecbfcd2a528c7502db192412c5a16R235 due to the lack of a better suggestion. I'm open for improvements. I certainly looks like too much boilerplate.